extra/mariabackup/xtrabackup.cc

```diff
@@ -4708,38 +4727,70 @@ static void stop_backup_threads()
   mysql_cond_destroy(&log_copying_stop);
 }
 
-/** Implement the core of --backup
-@return	whether the operation succeeded */
-bool Backup_datasinks::backup_low()
+/**
+Wait for enough log to be copied.
+@return whether log_copying_thread() copied everything until the target lsn
+*/
+static bool backup_wait_for_commit_lsn()
 {
-	mysql_mutex_lock(&recv_sys.mutex);
-	ut_ad(!metadata_to_lsn);
+  lsn_t lsn= get_current_lsn(mysql_connection);
+  mysql_mutex_lock(&recv_sys.mutex);
+  ut_ad(!metadata_to_lsn);
+  ut_ad(!metadata_last_lsn);
 
-	/* read the latest checkpoint lsn */
-	{
-		const lsn_t lsn = recv_sys.lsn;
-		if (recv_sys.find_checkpoint() == DB_SUCCESS
-		    && log_sys.is_latest()) {
-			metadata_to_lsn = log_sys.next_checkpoint_lsn;
-			msg("mariabackup: The latest check point"
-			    " (for incremental): '" LSN_PF "'",
-			    metadata_to_lsn);
-		} else {
-			msg("Error: recv_sys.find_checkpoint() failed.");
-		}
+  lsn_t last_lsn= recv_sys.lsn;
 
-		recv_sys.lsn = lsn;
-		stop_backup_threads();
-	}
+  /* read the latest checkpoint lsn */
+  {
+    if (recv_sys.find_checkpoint() == DB_SUCCESS && log_sys.is_latest())
+    {
+      if (log_sys.next_checkpoint_lsn > lsn)
+        lsn= log_sys.next_checkpoint_lsn;
+      metadata_to_lsn= log_sys.next_checkpoint_lsn;
+      msg("mariabackup: The latest check point (for incremental): '"
+          LSN_PF "'", metadata_to_lsn);
+    }
+    else
+    {
+      msg("Error: recv_sys.find_checkpoint() failed.");
+      stop_backup_threads();
+      mysql_mutex_unlock(&recv_sys.mutex);
+      return false;
+    }
 
-	if (metadata_to_lsn && xtrabackup_copy_logfile()) {
-		mysql_mutex_unlock(&recv_sys.mutex);
-		ds_close(dst_log_file);
-		dst_log_file = NULL;
-		return false;
-	}
+    recv_sys.lsn= last_lsn;
+  }
 
-	mysql_mutex_unlock(&recv_sys.mutex);
+  ut_ad(metadata_to_lsn);
+  msg("Waiting for log copy thread to read lsn " LSN_PF, lsn);
+  metadata_last_lsn= lsn;
+
+  while (log_copying_running)
+  {
+    timespec abstime;
+    set_timespec(abstime, 5);
+    if (my_cond_timedwait(&scanned_lsn_cond, &recv_sys.mutex.m_mutex,
+                          &abstime) &&
+        last_lsn == recv_sys.lsn)
+      break;
+    last_lsn= recv_sys.lsn;
+  }
+
+  metadata_last_lsn= last_lsn= recv_sys.lsn;
+  stop_backup_threads();
+  mysql_mutex_unlock(&recv_sys.mutex);
+  return backup_wait_timeout(lsn, last_lsn);
```

Comment:
I refactored the `while` loop to a separate function `backup_wait_for_lsn_low()`. The condition `while (log_copying_running && last_lsn < lsn)` can be modified to what we need here by passing `lsn=LSN_MAX` so that we will actually wait until `!log_copying_running`.